### DISCUSSION TITLE
Implicit `ToString` call on method parameters of `string` type

### DISCUSSION THREAD

**Discussion by tvardero** [dotnet/csharplang#6857] [Language Ideas] [answered]
QoL / syntax sugar.

## Current behavior (shown on example):
If you pass variable not of string type to method, that expects string parameter - you will receive CS1503 compiler error: "unable to convert ... to type string".

```csharp
// System.Security.Claims.dll
public class ClaimsPrincipal : IPrincipal
{
    // ...
    public virtual bool HasClaim(string type, string value);
    // ...
}

// MyProject.dll
public class FooClaimType
{
    public required string Name { get; init; }
    public string? Description { get; init; }
    public string[] AllowedValues { get; init; } = Array.Empty<string>();

    public static explicit operator string(FooClaimType obj) => obj.Name;
}
```

```csharp
var isBarClaim = new FooClaimType { Name = "is_bar" };

// The code below results in CS1503 compiler error: 
// "unable to convert FooClaimType to type string" 
// and "unable to convert bool to type string"
var isBar = principal.HasClaim(isBarClaim, true); 
```

## Proposed:
Add marker words `implicit` and `explicit` for string type parameters.
Add `<ImplicitToStringConversion>` property to `<PropertyGroup>` of csproj / fsproj / vbproj files, to define default behavior project-wide.

Explicit results in CS1503 compiler error, if type of the passed argument is not `string`.
Implicit instead converts the passed argument of any type to string using `implicit operator string`, then `explicit operator string`, on last resort `override ToString()`.

Current behavior is "explicit".

```csharp
// System.Security.Claims.dll
public class ClaimsPrincipal : IPrincipal
{
    // ...
    public virtual bool HasClaim(implicit string type, implicit string value);
    // ...
}
```

```csharp
var isBarClaim = new FooClaimType { Name = "is_bar" };

var isBar = principal.HasClaim(isBarClaim, true);
// is converted by compiler to 
var isBarCompiled = principal.HasClaim((string)isBarClaim, true.ToString());
```

### Motivation:
Less visual noise on methods calls. Especially when method has a large amount of string parameters (as an example, some methods of EF MigrationBuilder).

**Comment by HaloFour** [comment, marked as answer]
[Extension everything](https://github.com/dotnet/csharplang/issues/192) might allow you to define your own implicit string conversion on any given class.  Beyond that I don't think this should be a language feature as it seemingly encourages the wrong behavior and, in my opinion, that "visual noise" is clarity that shouldn't be eliminated.
